### DISCUSSION TITLE
Elasticsearch Migrations

### DISCUSSION THREAD

**Discussion by mrflip** [theory/sqitch#638] [Q&A]
We use a postgres database for serving and writes, pushing data to Elastic for search. We're looking at using Sqitch for the postgres part, and it would be convenient to have Sqitch manage both. Some questions:

Q 0. First, a meta-question for people with experience in ETL -- given that the ES database is intended to always mirror the contents of PG, is it a better idea to NOT perform in-place migrations on the downstream store, but manage things by crossloading new copies of the downstream database and then shifting traffic?

Q1. what is the learning curve like for adding a new DB target to sqitch? It's two decades since I last used Perl, but I'd muddle through; more worried about having to reach into complicated parts of the code.

Q2. Would it matter that the api is not SQL? ES's api is beautiful and straightforward, so I'd think to do it by either
 - writing scripts as the query in JSON, and use either perl's ES client or just plain REST to send it across; or
 - shelling out to eg. the Elastic/Curator command-line tool. 
 Is it already possible to have migration steps be shell scripts?

Q3. ES is weird and wonderful and I love it but some types of evolution are still a horrorshow -- some changes to a column's "mapping" can require a full table reindex. This can be done with no downtime, and even quickly if you set up aliases correctly, but there's just a lot more drama surrounding the affair. 

Point being, it may be that one would want the flow to stop at some point to allow some non-sqitch process to execute the migration, and to have it pick up at the verify part of that step. Is that supported?

**Comment by theory** [comment]
> Q 0. First, a meta-question for people with experience in ETL -- given that the ES database is intended to always mirror the contents of PG, is it a better idea to NOT perform in-place migrations on the downstream store, but manage things by crossloading new copies of the downstream database and then shifting traffic?

I don't know, frankly. Pg and ES are radically different databases. In a configuration at a past job, we had our service publish to a message queue, and a consumer read from it and kept Elasticsearch up-to-date. You could do the same here, something like using [LISTEN/NOTIFY](https://www.postgresql.org/docs/14/libpq-notify.html) to publish changes right from the database. I don't know to what extent you'd need to keep changes in sync, like adding or removing a column, but from what I understand Elasticsearch can be set up to take a dynamic list of indexes and values.

> Q1. what is the learning curve like for adding a new DB target to sqitch? It's two decades since I last used Perl, but I'd muddle through; more worried about having to reach into complicated parts of the code.

You shouldn't have to do any Perl at all. Especially with Postgres, where installing stuff is pretty easy. Or just use the [docker image](https://github.com/sqitchers/docker-sqitch), which contains `psql` and all the required libraries for Postgres database management.

> Q2. Would it matter that the api is not SQL? ES's api is beautiful and straightforward, so I'd think to do it by either

Okay now your'e getting into needing to re-learn Perl to some degree, because as-of today there is no support for Elasticsearch. But understand that fundamental to the Sqitch philosophy is to rely on the command-line scripting tools that come with each database, like `psql` for Postgres and SQL\*Plus for Oracle. Does Elastic have a CLI for passing DDL files to? I would expect you'd use the API and the Perl Elastic client to maintain the registry table, of course.

> Q3. ES is weird and wonderful and I love it but some types of evolution are still a horrorshow -- some changes to a column's "mapping" can require a full table reindex. This can be done with no downtime, and even quickly if you set up aliases correctly, but there's just a lot more drama surrounding the affair.
>
> Point being, it may be that one would want the flow to stop at some point to allow some non-sqitch process to execute the migration, and to have it pick up at the verify part of that step. Is that supported?
 
Sqitch shells out to the database client to run the deploy, verify, and revert scripts, each as a separate invocation. But all the existing engines are RDBMSes that use SQL; it may well be that you end up needing to do something different to implement the Elastic engine.

HTH!
